File: operator/src/controller.rs
//! Reconciliation controller for SsdkvCluster resources.

use std::sync::Arc;
use std::time::Duration;

use k8s_openapi::api::apps::v1::StatefulSet;
use k8s_openapi::api::core::v1::{ConfigMap, Service};
use kube::api::{Api, Patch, PatchParams};
use kube::runtime::controller::Action;
use kube::{Client, CustomResource, ResourceExt};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tracing::{error, info};

use crate::resources;
use crate::shard_manager;

/// The spec for an SsdkvCluster resource.
#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[kube(
    group = "ssdkv.io",
    version = "v1alpha1",
    kind = "SsdkvCluster",
    namespaced,
    status = "SsdkvClusterStatus"
)]
#[serde(rename_all = "camelCase")]
pub struct SsdkvClusterSpec {
    /// Number of nodes in the cluster.
    #[serde(default = "default_replicas")]
    pub replicas: i32,

    /// Replication factor.
    #[serde(default = "default_replication_factor")]
    pub replication_factor: i32,

    /// Storage configuration.
    #[serde(default)]
    pub storage: StorageSpec,

    /// Container image.
    #[serde(default = "default_image")]
    pub image: String,

    /// Redis-compatible port.
    #[serde(default = "default_redis_port")]
    pub redis_port: i32,

    /// Cluster communication port.
    #[serde(default = "default_cluster_port")]
    pub cluster_port: i32,

    /// Resource requests/limits.
    #[serde(default)]
    pub resources: Option<ResourceSpec>,

    /// Allow replica nodes to serve read requests (READONLY command).
    #[serde(default)]
    pub replica_read: bool,
}

fn default_replicas() -> i32 { 3 }
fn default_replication_factor() -> i32 { 2 }
fn default_image() -> String { "ssd-kv:latest".to_string() }
fn default_redis_port() -> i32 { 7777 }
fn default_cluster_port() -> i32 { 7780 }

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema, Default)]
#[serde(rename_all = "camelCase")]
pub struct StorageSpec {
    #[serde(default = "default_storage_size")]
    pub size: String,
    pub storage_class: Option<String>,
}

fn default_storage_size() -> String { "10Gi".to_string() }

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct ResourceSpec {
    pub requests: Option<ResourceValues>,
    pub limits: Option<ResourceValues>,
}

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct ResourceValues {
    pub cpu: Option<String>,
    pub memory: Option<String>,
}

/// Status subresource for SsdkvCluster.
#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema, Default)]
#[serde(rename_all = "camelCase")]
pub struct SsdkvClusterStatus {
    pub phase: Option<String>,
    pub ready_replicas: Option<i32>,
    pub topology_version: Option<i64>,
}

/// Shared context for the controller.
pub struct Context {
    pub client: Client,
}

/// Main reconciliation function.
pub async fn reconcile(
    cluster: Arc<SsdkvCluster>,
    ctx: Arc<Context>,
) -> Result<Action, kube::Error> {
    let name = cluster.name_any();
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    info!("Reconciling SsdkvCluster {}/{}", namespace, name);

    let client = &ctx.client;

    // 1. Create or update headless Service
    let service = resources::build_headless_service(&cluster);
    let svc_api: Api<Service> = Api::namespaced(client.clone(), &namespace);
    svc_api
        .patch(
            &format!("{}-headless", name),
            &PatchParams::apply("ssdkv-operator"),
            &Patch::Apply(service),
        )
        .await?;
    info!("Headless service ensured for {}", name);

    // 2. Create or update client Service
    let client_service = resources::build_client_service(&cluster);
    svc_api
        .patch(
            &name,
            &PatchParams::apply("ssdkv-operator"),
            &Patch::Apply(client_service),
        )
        .await?;
    info!("Client service ensured for {}", name);

    // 2b. Create or update readonly Service when replica_read is enabled
    if cluster.spec.replica_read {
        let readonly_service = resources::build_readonly_service(&cluster);
        svc_api
            .patch(
                &format!("{}-readonly", name),
                &PatchParams::apply("ssdkv-operator"),
                &Patch::Apply(readonly_service),
            )
            .await?;
        info!("Readonly service ensured for {}", name);
    }

    // 3. Compute shard assignments
    let shard_map = shard_manager::compute_shard_map(
        cluster.spec.replicas as u32,
        cluster.spec.replication_factor as u32,
    );

    // 4. Create or update topology ConfigMap
    let config_map = resources::build_topology_configmap(&cluster, &shard_map);
    let cm_api: Api<ConfigMap> = Api::namespaced(client.clone(), &namespace);
    cm_api
        .patch(
            &format!("{}-topology", name),
            &PatchParams::apply("ssdkv-operator"),
            &Patch::Apply(config_map),
        )
        .await?;
    info!("Topology configmap ensured for {}", name);

    // 5. Create or update StatefulSet
    let statefulset = resources::build_statefulset(&cluster);
    let ss_api: Api<StatefulSet> = Api::namespaced(client.clone(), &namespace);
    ss_api
        .patch(
            &name,
            &PatchParams::apply("ssdkv-operator"),
            &Patch::Apply(statefulset),
        )
        .await?;
    info!("StatefulSet ensured for {}", name);

    // 6. Update status based on actual StatefulSet state
    let ss_status = ss_api.get(&name).await?;
    let actual_ready = ss_status
        .status
        .as_ref()
        .and_then(|s| s.ready_replicas)
        .unwrap_or(0);
    let desired = cluster.spec.replicas;
    let phase = if actual_ready == desired {
        "Running"
    } else if actual_ready > 0 {
        "Degraded"
    } else {
        "Pending"
    };
    let status = SsdkvClusterStatus {
        phase: Some(phase.to_string()),
        ready_replicas: Some(actual_ready),
        topology_version: Some(1),
    };

    let status_patch = serde_json::json!({
        "apiVersion": "ssdkv.io/v1alpha1",
        "kind": "SsdkvCluster",
        "status": status,
    });

    let cluster_api: Api<SsdkvCluster> = Api::namespaced(client.clone(), &namespace);
    cluster_api
        .patch_status(
            &name,
            &PatchParams::apply("ssdkv-operator"),
            &Patch::Merge(status_patch),
        )
        .await?;

    info!("Reconciliation complete for {}/{}", namespace, name);

    // Requeue after 60 seconds for periodic health check
    Ok(Action::requeue(Duration::from_secs(60)))
}

/// Error policy: log and retry with backoff.
pub fn error_policy(
    _cluster: Arc<SsdkvCluster>,
    err: &kube::Error,
    _ctx: Arc<Context>,
) -> Action {
    error!("Reconcile error: {:?}", err);
    Action::requeue(Duration::from_secs(30))
}

File: operator/controller.go
package main

import (
	"context"
	"fmt"

	appsv1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/types"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
	"sigs.k8s.io/controller-runtime/pkg/log"
)

// SsdkvClusterReconciler reconciles SsdkvCluster objects.
type SsdkvClusterReconciler struct {
	client.Client
	Scheme *runtime.Scheme
}

func (r *SsdkvClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	logger := log.FromContext(ctx)

	// Fetch the SsdkvCluster resource
	var cluster SsdkvCluster
	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
		if errors.IsNotFound(err) {
			logger.Info("SsdkvCluster deleted", "name", req.NamespacedName)
			return ctrl.Result{}, nil
		}
		return ctrl.Result{}, err
	}

	logger.Info("Reconciling SsdkvCluster", "name", cluster.Name, "replicas", cluster.Spec.GetReplicas())

	// 1. Headless Service
	headlessSvc := buildHeadlessService(&cluster)
	if err := r.reconcileService(ctx, &cluster, headlessSvc); err != nil {
		return ctrl.Result{}, fmt.Errorf("headless service: %w", err)
	}

	// 2. Client Service
	clientSvc := buildClientService(&cluster)
	if err := r.reconcileService(ctx, &cluster, clientSvc); err != nil {
		return ctrl.Result{}, fmt.Errorf("client service: %w", err)
	}

	// 3. StatefulSet
	desiredSS, err := buildStatefulSet(&cluster)
	if err != nil {
		cluster.Status.Phase = "Invalid"
		_ = r.Status().Update(ctx, &cluster)
		return ctrl.Result{}, fmt.Errorf("build statefulset: %w", err)
	}
	if err := r.reconcileStatefulSet(ctx, &cluster, desiredSS); err != nil {
		return ctrl.Result{}, fmt.Errorf("statefulset: %w", err)
	}

	// 4. Update status
	var currentSS appsv1.StatefulSet
	if err := r.Get(ctx, types.NamespacedName{Name: cluster.Name, Namespace: cluster.Namespace}, &currentSS); err == nil {
		cluster.Status.ReadyReplicas = currentSS.Status.ReadyReplicas
		if currentSS.Status.ReadyReplicas == cluster.Spec.GetReplicas() {
			cluster.Status.Phase = "Running"
		} else if currentSS.Status.ReadyReplicas > 0 {
			cluster.Status.Phase = "Degraded"
		} else {
			cluster.Status.Phase = "Pending"
		}
	}

	return ctrl.Result{}, nil
}

func (r *SsdkvClusterReconciler) reconcileService(ctx context.Context, owner *SsdkvCluster, desired *corev1.Service) error {
	if err := controllerutil.SetControllerReference(owner, desired, r.Scheme); err != nil {
		return err
	}

	var existing corev1.Service
	err := r.Get(ctx, types.NamespacedName{Name: desired.Name, Namespace: desired.Namespace}, &existing)
	if errors.IsNotFound(err) {
		return r.Create(ctx, desired)
	}
	if err != nil {
		return err
	}

	// Update ports (preserve ClusterIP)
	existing.Spec.Ports = desired.Spec.Ports
	existing.Spec.Selector = desired.Spec.Selector
	return r.Update(ctx, &existing)
}

func (r *SsdkvClusterReconciler) reconcileStatefulSet(ctx context.Context, owner *SsdkvCluster, desired *appsv1.StatefulSet) error {
	if err := controllerutil.SetControllerReference(owner, desired, r.Scheme); err != nil {
		return err
	}

	var existing appsv1.StatefulSet
	err := r.Get(ctx, types.NamespacedName{Name: desired.Name, Namespace: desired.Namespace}, &existing)
	if errors.IsNotFound(err) {
		return r.Create(ctx, desired)
	}
	if err != nil {
		return err
	}

	// Update mutable fields
	existing.Spec.Replicas = desired.Spec.Replicas
	existing.Spec.Template = desired.Spec.Template
	return r.Update(ctx, &existing)
}

// SetupWithManager registers the controller with the manager.
func (r *SsdkvClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
	// Register CRD types with scheme
	schemeBuilder := runtime.NewSchemeBuilder(func(scheme *runtime.Scheme) error {
		scheme.AddKnownTypes(
			SchemeGroupVersion,
			&SsdkvCluster{},
			&SsdkvClusterList{},
		)
		metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
		return nil
	})
	if err := schemeBuilder.AddToScheme(mgr.GetScheme()); err != nil {
		return err
	}

	return ctrl.NewControllerManagedBy(mgr).
		For(&SsdkvCluster{}).
		Owns(&appsv1.StatefulSet{}).
		Owns(&corev1.Service{}).
		Complete(r)
}
